Add sanity check of tar "uid, "gid" and "mtime" fields

josusky opened the pull request:
Sometimes a bz2 file is identified as tar.bz2, i.e. a chunk of
raw data may look like a tar header (metadata block).
To reduce chance of such false positive, the format of uid, gid
and mtime field is checked. This fields are expected to contain
octal numbers. In fact, also space and '\0' are allowed - perhaps
I am too liberal.

kientzle commented:
Can the UID or GID fields ever contain base-256 values?

josusky commented:
In theory it shouldn't.

But I would also like to have it confirmed by practical experiences of  
people who use some less common systems, let say old Solaris :-)

Here is example of tar in v7 format where the octal GID/UID/mtime (and few  
other fields) are clearly visible:
hexdump -C test-v7.tar
00000000 68 75 68 75 00 00 00 00 00 00 00 00 00 00 00 00 |huhu............|
00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |................|
*
00000060 00 00 00 00 30 30 30 30 36 36 30 00 30 30 30 31 |....0000660.0001|
00000070 37 35 30 00 30 30 30 31 37 35 30 00 30 30 30 30 |750.0001750.0000|
00000080 30 30 30 30 30 30 30 00 31 33 30 31 34 30 30 35 |0000000.13014005|
00000090 33 37 37 00 30 30 36 36 35 37 00 20 00 00 00 00 |377.006657. ....|
000000a0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |................|
*
00000140 00 00 00 00 00 00 00 00 00 30 30 30 30 30 30 30 |.........0000000|
00000150 00 30 30 30 30 30 30 30 00 00 00 00 00 00 00 00 |.0000000........|
00000160 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |................|

("huhu" is name of the test file, "1750" is my GID a UID in octal)


On Sat, 19 Nov 2016 08:25:20 +0100, Tim Kientzle  
<notifications@github.com> wrote:

>
> Can the UID or GID fields ever contain base-256 values?

kientzle commented:
Modern POSIX systems typically allow 32-bit UIDs and GIDs, which can overflow the 8-byte fields here when expressed in octal.  So you should permit a base-256 value.  A more complete check would then be:
* If the first byte is zero (base-256 marker), then all but the last four bytes must be zero (the last four bytes can be anything).
* Otherwise, all bytes must be space, octal, or null.

josusky commented:
You were right. It can be base256. The POSIX format is fine, the new version keeps the old fields in proper format but the GNU tar implementation does nasty tricks. I adjusted the sanity check and added few tests.

josusky commented:
Hm, I was not able to create such tar file neither with gnu tar nor with  
bsdtar - but it probably proves only that I cannot read manuals. I forged  
such file with hexeditor, to have something to test on, and adjusted the  
validation function as you suggested.

On Tue, 22 Nov 2016 21:31:42 +0100, Tim Kientzle  
<notifications@github.com> wrote:

>
> @kientzle commented on this pull request.
>
> In libarchive/archive_read_support_format_tar.c:
>> @@ -294,6 +294,36 @@ archive_read_format_tar_cleanup(struct  
>> archive_read *a)
> 	return (ARCHIVE_OK);
> }
>+static int
> +validate_number_field(const char* p_field, size_t i_size)
> +{
> +	unsigned char marker = (unsigned char)p_field[0];
> +	/* octal? */
> +	if ((marker >= '0' && marker <= '7') || marker == ' ') {
> +		/* must be terminated by null or space */
> +		if (p_field[i_size - 1] != '\0' && p_field[i_size - 1] != ' ') {
> +			return 0;
> +		}
>
> This looks much better, but is still a little too strict: POSIX  
> currently insists that fields be terminated with >null or space, but  
> those rules have changed over the years, and some implementations  
> (including libarchive itself) >will omit the terminator if they need to  
> fit a large value. This implies the last character could be an octal  
> digit.
>
> I suggest structuring this check as follows:
> initialize i to 0.
> Skip over any leading spaces (until end of field or non-space)
> If end of field, fail.
> Next character must be octal.
> Skip over any remaining octal characters (until end of field or  
> non-octal)
> If not end of field, next character must be space or null.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

kientzle commented:
Thanks for your hard work on this!
